client/modules/ftml-wasm-worker/src/worker/ftml.worker.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * @file Web-workerized version of "ftml-wasm".
+ */
+
+import * as FTML from "ftml-wasm"
+import { expose, transfer, decode } from "./lib"
+
+const ready = FTML.loading
+
+expose({
+  async init(wasmURL: string) {
+    await FTML.init(wasmURL)
+    return true
+  },
+
+  async version() {
+    await ready
+    return transfer(FTML.version())
+  },
+
+  async preprocess(raw: ArrayBuffer) {
+    await ready
+    const str = decode(raw)
+    return transfer(FTML.preprocess(str))
+  },
+
+  async tokenize(raw: ArrayBuffer) {
+    await ready
+    const str = decode(raw)
+    return FTML.tokenize(str)
+  },
+
+  async parse(raw: ArrayBuffer) {
+    await ready
+    const str = decode(raw)
+    return FTML.parse(str)
+  },
+
+  async renderHTML(raw: ArrayBuffer) {
+    await ready
+    const str = decode(raw)
+    const { html } = FTML.render(str)
+    return transfer(html)
+  },
+
+  async renderStyle(raw: ArrayBuffer) {
```

Comment:
What is this? Doing the rendering but only extracting the CSS?

I think I've seen this elsewhere, why can you run the render once and then use both the HTML and CSS?

Reply:
I'll have to see how it plays out - but the idea is that when you're just transferring strings back and forth, you can use `ArrayBuffer` encoded strings to transfer between the web-worker and the main thread with very low latency and performance cost.

To explain a bit further, when you "return" an object to the main thread, or the other way around, the browser has to do something called a "structured clone", which duplicates the entire object. When you "transfer" an `ArrayBuffer`, that doesn't happen.

Considering that for SCP-1730 the HTML is ~220KB, I don't want to be cloning objects constantly that are of that size, or larger.

Reply:
Makes sense.

But would it be possible to have two `ArrayBuffer`s, one for HTML and one for CSS? Then you only do a single request and get both bits of data.

Reply:
Maybe? Honestly, I haven't thought about it. I'm not sure if you can do that in one "return", at least with `threads`.

(`threads` is the NPM package I'm using to abstract web-worker handling - it's very nice for handling a lot of the agonizing chores you have to do.)

Reply:
![image](https://user-images.githubusercontent.com/34875062/115476781-cfcd4200-a1ff-11eb-9fa1-47e4e9738da6.png)

Oops. Yeah, you can. Let me add that.

Reply:
![image](https://user-images.githubusercontent.com/34875062/115477934-423f2180-a202-11eb-8776-6f39355f98ff.png)

There we go.